docs/accesscontrol.1
.\" Automatically generated by Pandoc 2.17.1.1
.\"
.\" Define V font for inline verbatim, using C font in formats
.\" that render this, and otherwise B font.
.ie "\f[CB]x\f[]"x" \{\
. ftr V B
. ftr VI BI
. ftr VB B
. ftr VBI BI
.\}
.el \{\
. ftr V CR
. ftr VI CI
. ftr VB CB
. ftr VBI CBI
.\}
.TH "ACCESS-CONTROL" "1" "20 Jun 2023" "v0.1" "Tool for testing access control - documentation"
.hy
.SH NAME
.PP
accesscontrol - script to test access control on a web app during a
pentest
.SH SYNOPSIS
.TP
accesscontrol [-h] [-c CONFIG] [\[en]login-path LOGIN_PATH] [\[en]id-field ID_FIELD]
.nf
\f[C]
       [--pwd-field PWD_FIELD] [--login-code LOGIN_CODE | --login-text LOGIN_TEXT] [--csrf]
\f[R]
.fi
.nf
\f[C]
       [--csrf-class CSRF_CLASS] [--csrf-name CSRF_NAME] [-o OUT] [-w WAIT]
\f[R]
.fi
.nf
\f[C]
       [--limit-users LIMIT_USERS] [--limit-urls LIMIT_URLS] [-m MAX_RETRIES] [-t TIMEOUT] [-p] [-v] [-d] [-j]
\f[R]
.fi
.nf
\f[C]
       [--allow-redirect] [--disable-unauth] [--disable-https]
\f[R]
.fi
.nf
\f[C]
       [host] [users] [urls]
\f[R]
.fi
.SH DESCRIPTION
.PP
With a given list of users and URLs and a host server, test all accesses
and store result as a table or in JSON format.
For the moment, this tool return only the HTTP status code of the
response.
Arguments can be specified from the command line or from a configuration
file (see: FILES).
.PP
The script will first send a GET request to all the URLs in the given
file without any authentication (unless \f[B]disable-unauth\f[R] is
set).
Then it will try to login to the application with the given parameters
(\f[B]login-path\f[R], \f[B]id-field\f[R], \f[B]pwd-field\f[R],
\f[B]login-code\f[R], \f[B]login-text\f[R]).
To authenticate the script send a POST request to the page given in
\f[B]login-path\f[R].
The payload is created with the \f[B]id-field\f[R] and
\f[B]pwd-field\f[R] parameters and with the values parsed from the CSV
file containing the users credentials (\f[B]users\f[R] parameters).
.PP
It\[cq]s possible to use a CSRF mechanism to connect as a classic user
do.
You must enable CSRF using with \f[B]csrf\f[R] parameter and then, you
can give the type of HTML tag which contains the CSRF token with
\f[B]csrf-class\f[R] and their name with \f[B]csrf-name\f[R].
.PP
\f[B]Example:\f[R]
.PP
With default parameters (see: OPTIONS), the following request will be
send for logging in:
.IP
.nf
\f[C]
POST /login HTTP/1.1
[... Headers ...]

email=user1&password=password1
\f[R]
.fi
.PP
And then, by default the script detect a valid login attempt with a HTTP
200 response, but you can change the status code, or detect a valid
authentication with a string in the response text.
By default, the tool does not follow redirections, for example, to
detect a valid authentication upon a 302 status code in HTTP response,
but you can enable the redirection.
.SH OPTIONS
.TP
-h, \[en]help
show help message and exit
.SS Load arguments from configuration file:
.TP
-c CONFIG, \[en]config CONFIG
configuration file to load arguments from, JSON format, override command
line arguments
.SS Required arguments without configuration file:
.TP
host
specify the target host
.TP
users
file containing the list of users
.TP
urls
file containing the list of URLs
.SS Login helper arguments:
.TP
\[en]login-path LOGIN_PATH
path to the login page of the application (default: /login)
.TP
\[en]id-field ID_FIELD
name of the id field in the request for login (default: email)
.TP
\[en]pwd-field PWD_FIELD
name of the password field in the request for login (default: password)
.TP
\[en]login-code LOGIN_CODE
status code upon a valid login request (default: 200)
.TP
\[en]login-text LOGIN_TEXT
text in the page upon a valid upon login request
.TP
\[en]csrf
enable use of a CSRF token for login (default: False)
.TP
\[en]csrf-class CSRF_CLASS
class of the CSRF token element in the login page (default: input)
.TP
\[en]csrf-name CSRF_NAME
name of the CSRF token element in the login page (default: csrf)
.SS Optional arguments:
.TP
-o OUT, \[en]out OUT
store the output in this file
.TP
-w WAIT, \[en]wait WAIT
waiting time to limit requests rate, in seconds (default: 1)
.TP
\[en]limit-users LIMIT_USERS
set the limit of users to test
.TP
\[en]limit-urls LIMIT_URLS
set the limit of URLs to test
.TP
-t TIMEOUT, \[en]timeout TIMEOUT
timeout for each HTTP request (default: 5)
.TP
-m MAX_RETRIES, \[en]max-retries MAX_RETRIES
max retries attempts if request failed (default: 3)
.TP
-p, \[en]proxy
enable use of proxy, for the moment: 127.0.0.1:8080 (default: False)
.TP
-v, \[en]verbose
see info and warning messages (default: False)
.TP
-d, \[en]debug
see debug messages (default: False)
.TP
-j, \[en]json
save results in JSON format (default: False)
.TP
\[en]allow-redirect
follow requests redirection (default: False)
.TP
\[en]disable-unauth
disable unauthenticated tests (default: False)
.TP
\[en]disable-https
disable HTTPS (default: False)
.SH FILES
.SS Users file
.PP
The file containing the users credentials must fit the following format,
including \[lq]\f[V];\f[R]\[rq] delimiter:
.IP
.nf
\f[C]
role;username;password
user;johndoe;Password#123
administrator;admin;P\[at]ssw0rd456
\f[R]
.fi
.SS Configuration file
.PP
Parameters can be load from a configuration file.
Here is an example of this file :
.IP
.nf
\f[C]
{
  \[dq]host\[dq]: \[dq]example.domain.fr\[dq],
  \[dq]users\[dq]: \[dq]user.csv\[dq],
  \[dq]urls\[dq]: \[dq]url.txt\[dq],
  \[dq]login_path\[dq]: \[dq]/login\[dq],
  \[dq]id_field\[dq]: \[dq]email\[dq],
  \[dq]pwd_field\[dq]: \[dq]password\[dq],
  \[dq]login_code\[dq]: 200,
  \[dq]login_text\[dq]: null,
  \[dq]csrf\[dq]: false,
  \[dq]csrf_class\[dq]: \[dq]input\[dq],
  \[dq]csrf_name\[dq]: \[dq]csrf\[dq],
  \[dq]out\[dq]: null,
  \[dq]wait\[dq]: 1,
  \[dq]limit_users\[dq]: null,
  \[dq]limit_urls\[dq]: null,
  \[dq]timeout\[dq]: 5,
  \[dq]max_retries\[dq]: 3,
  \[dq]proxy\[dq]: false,
  \[dq]verbose\[dq]: false,
  \[dq]debug\[dq]: false,
  \[dq]json\[dq]: false,
  \[dq]allow_redirect\[dq]: false,
  \[dq]disable_unauth\[dq]: false,
  \[dq]disable_https\[dq]: false
}
\f[R]
.fi
.SH TO DO
.IP \[bu] 2
Add possibility of test control access with POST requests with
parameters
.IP \[bu] 2
Possibility of adding headers to the requests
.IP \[bu] 2
Auto detection of login method (in progress)
.SH AUTHORS
Louka.


source/accesscontrol.py
#!/usr/bin/python3

from urllib.parse import urlparse
from json import dump, load, dumps
from csv import DictReader
from time import sleep

import logging
import source.usersession as us

class AccessControl():
    """
    This class implements the main access control test. It loads users and URLs from given files, login
    and execute a GET request to each URL with each given account. Then results are saved or printed to
    standard output.
    """

    # Constants
    UNAUTHENTICATED = "Unauth. user"
    # Default proxy
    DEFAULT_PROXY = {
        "http":  "127.0.0.1:8080",
        "https": "127.0.0.1:8080"
    }
    # Users CSV fields name
    desc_field = "role"
    user_field = "username"
    pwd_field  = "password"
    
    def __init__(self, args_dict: dict):
        self.__dict__.update(args_dict)
        self.proxy = self.DEFAULT_PROXY if self.proxy else None
        # Logger paramters
        log_level = loggin.DEBUG if self.debug else (logging.INFO if self.verbose else logging.WARNING)
        logging.basicConfig(level=log_level)
        self.log = logging.getLogger("AccessControl")

    def loadURLs(self) -> list:
        """Load all URLs from file and keeps only the path part of them."""
        urls_file = self.urls
        with open(urls_file, 'r') as urlf:
            # read url list, split by line and remove last element (empty)
            content = urlf.read().split('\n')[:-1]
        urls = list()
        for url in content:
            p = urlparse(url)
            urls.append(p.path + ('?' + p.query if p.query != '' else ''))
        return urls

    def loadUsers(self) -> list:
        """
        Load all users from the given CSV file.
        The format must fit the described format in the documentation
        """
        users_file = self.users
        users = list()
        with open(users_file, 'r') as userf:
            user_reader = DictReader(userf, delimiter=';')
            # load users (list of dictionnaries)
            for user in user_reader:
                users.append(user)
        return users

    def save(self, result: dict):
        """Save result dict to the given output file or print it to stdout, as JSON or table format"""
        output = sys.stdout
        # open output file
        if self.out is not None:
            try:
                output = open(self.out, 'w')
            except:
                log.warn("Can't open given output file, write to stdout")
        # Save in JSON format
        if self.json:
            dump(result, output, indent=4)
            if output == sys.stdout:
                print()
        # Save in human-readable format
        else:
            self.writeTable(result, output)
        # Close output file if opened
        if self.out is not None:
            output.close()

    def writeTable(self, data: dict, out):
        """Write the result dict as a table to the output"""
        # determine padding
        max_pad = max([len(path) for path in data.keys()])
        users = list(data.values())[0].keys()
        column_length = max([len(u) for u in users]) + 1
        # print users
        fmt_str = ("{:<%d} " % max_pad) + ("{:<%d}" % column_length) * len(users) + '\n'
        out.write(fmt_str.format("URLs", *users))
        out.write('-' * (max_pad + 1 + column_length * len(users)) + '\n')
        # print table
        for path, codes in zip(data.keys(), data.values()):
            fmt_str = ("{:<%d} " % max_pad) + ("{:<%d}" % column_length) * len(codes) + '\n'
            out.write(fmtStr.format(path, *codes.values()))
 
    # Test all access controls with a GET request to each urls for each user
    def execRequests(self, users: list, urls: list) -> dict:
        """
        Main function of the class. It will login with each user credentials and then
        send a request to all URLs to get all the status codes. The data are returns
        in a dictionnary containing all informations. If a wait duration was given, it
        will wait between each account and after login to limit the request rate.
        """
        # init result
        result = dict()
        for url in urls:
            result[url] = dict()
        # Add unauthenticated user if needed
        if not self.disable_unauth:
            users = [{
                self.desc_field: self.UNAUTHENTICATED,
                self.user_field: None,
                self.pwd_field: None
            }] + users
        # Send all requests for each users
        for user in users:
            self.log.info("Testing %d URLs for user: %s, role: %s" % (len(urls), user[self.user_field], user[self.desc_field]))
            session = us.UserSession(self.host,
                self.login_path, self.id_field, self.pwd_field,
                self.login_code, (self.login_text is not None), self.login_text,
                self.csrf, self.csrf_name, self.csrf_class,
                self.disable_https, self.allow_redirect, self.timeout, self.proxy
            )
            # Try to login
            if user[self.desc_field] is not self.UNAUTHENTICATED:
                session.login(user[self.user_field], user[self.pwd_field])
                # Test if connection suceed
                if not session.isConnected:
                    log.warn("Skipping user %s: can't log in" % user[self.user_field])
                    continue
            # Add sleep time to reduce request rate
            sleep(self.wait)
            for url in urls:
                try:
                    r = session.get(url)
                    count = 0
                    while r.status_code == 502 and count < self.max_retries:
                        self.log.debug(f"Got 502 when trying to access {url}, retrying")
                        r = session.get(url)
                        count += 1
                        sleep(self.wait)
                    result[url][user[self.desc_field]] = r.status_code
                # request timeout
                except requests.exceptions.Timeout:
                    self.log.debug(f"Request to {url} timeout")
                    result[url][user[self.desc_field]] = "Timeout"
            sleep(self.wait * 3)
        return result

       
    def run(self):
        """
        Run the main test by loading users and URLs, setting up options, sending requests
        and finally save or print the result.
        """
        # Host informations
        self.log.info(f"Host: {self.host}")
        # Proxy informations
        if self.proxy is not None:
            self.log.info("HTTP proxy is  : %s" % self.proxy['http'])
            self.log.info("HTTPS proxy is : %s" % self.proxy['https'])
        try:
            # Load users
            self.log.info(f"Load users file: {self.users}")
            users = self.loadUsers()
            # Load URLs
            self.log.info(f"Load target URLs file: {self.urls}")
            log.warn("WARNING: ensure /logout or equivalent is not in URLs or at the end of the list")
            urls  = self.loadURLs()
        except:
            self.log.error("Can't load or open users or URLs file")
            exit(1)
        # Limits the numbers of users and URLs to test if asked (debug)
        if self.limit_users is not None:
            users = users[:self.limit_users]
        if self.limit_urls is not None:
            urls = urls[:self.limit_urls]
        # Start test
        result = self.execRequests(users, urls)
        self.save(result)
        # Finished
        self.log.info("Done")

